akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1088,38 +1094,50 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
 
     private def sendEvent(event: T, promise: Promise[Done]): Promise[Done] = {
       onAsyncInput(event, OptionVal.Some(promise))
-      currentState.get() match {
-        case Completed ⇒ failPromiseOnComplete(promise)
-        case _         ⇒ promise
-      }
+      if (stopped) failPromiseOnComplete(promise)
+      else promise
     }
 
     // external call
     override def invokeWithFeedback(event: T): Future[Done] = {
       val promise: Promise[Done] = Promise[Done]()
-      promise.future.andThen { case _ ⇒ waitingForProcessing.remove(promise) }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+      promise.future.onComplete(_ ⇒ onFeedbackCompleted(promise))(ExecutionContexts.sameThreadExecutionContext)
+
+      if (waitingForProcessing.isEmpty) {
+        asyncCallbacksInProgress.add(this)
```

Comment:
Isn't that mutation of `asyncCallbacksInProgress` unsafe because it will be invoked from an arbitrary thread in general? If we need synchronization anyway to mutate the set it could make sense to use an AtomicReference for the set and just `null` it out when the stage has been stopped instead of the extra volatile var.

Reply:
Yes it is, d'oh!